File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function(weights, days) {
    
    let min_capacity = -Infinity;
    let max_capacity = 0;
    for(let w of weights) {
        max_capacity += w;
        if( min_capacity < w ) {
            min_capacity = w;
        }
    }
    
    //console.log(min_capacity, max_capacity);
    
    let ispossible = (capacity) => {
        let d = 1, sum = 0;
        for(let i = 0; i < weights.length ; i++) {
            let w = weights[i];
            if(sum + w > capacity) {
                d++;
                sum = 0;
            }
            sum = sum + w;
            //console.log("days = " , d, sum)
        }
        //console.log(capacity, d,d <= days);
        return d <= days ? true : false;
    }
    let ans = max_capacity;
    let low = min_capacity, high = max_capacity;
    
    while(low < high) {
        let mid = low + parseInt( (high - low) / 2 );
        if(ispossible(mid)) {
            ans = Math.min(ans, mid);
            high = mid;
        } else {
            low = mid + 1;
        }
    }
    return ans;
};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public boolean isPossible(int[] weights, int days, int capacity) {
        int d = 1;
        int sum = 0;
        for(int wt : weights) {
            if(sum + wt > capacity) {
                d++;
                sum = 0;
            }
            sum += wt;
        }
        return d <= days ? true : false;
    }
    public int shipWithinDays(int[] weights, int days) {
        final int n = weights.length;
        int min_capacity = Integer.MIN_VALUE;
        int max_capacity = 0;
        for(int wt : weights) {
            max_capacity += wt;
            min_capacity = min_capacity < wt ? wt : min_capacity;
        }
        
        int low = min_capacity, high = max_capacity, ans = max_capacity;
        
        while(low <= high) {
            int capacity = low + ( (high - low) / 2 );
            if( isPossible(weights, days, capacity) == true) {
                ans = ans > capacity ? capacity : ans;
                high = capacity - 1;
            } else {
                low = capacity + 1;
            }
        }
        return ans;
        
    }
}